File: BelajarOOPCSharp/CSharpObjectOP/InheritEncap.cs
namespace GameOOP
{
	// Inheritance and Encapsulation
	public class Personel
	{
        private string name = "";
        private int health;
        private int attackPower;

        public string Name
		{
			get => name;
			set => name = value;
		}

        public int Health
        {
			get => health;
			set => health = value > 0 ? value : 0;
		}

		public int AttackPower
		{
			get => attackPower;
			set => attackPower = value > 0 ? value : 0;
		}

		public Personel(string name, int health, int attackPower)
		{
			Name = name;
			Health = health;
			AttackPower = attackPower;
		}

		public virtual void Attack(Personel target)
		{			
			target.Health -= AttackPower;
			Console.WriteLine($"{target.Name} has {target.Health} health left.");
		}
	}

	public class Ranger : Personel 
	{
		public Ranger(string name, int health, int attackPower) : base(name, health, attackPower)
		{
		}

		public override void Attack(Personel target)
		{
			Console.WriteLine($"{Name} shoots an arrow at {target.Name} for {AttackPower} damage!");
			base.Attack(target);
		}
	}

	public class Villain : Personel
	{
		public Villain(string name, int health, int attackPower) : base(name, health, attackPower)
		{
		}
		public override void Attack(Personel target)
		{
			Console.WriteLine($"{Name} casts a dark spell on {target.Name} for {AttackPower} damage!");
			base.Attack(target);
		}
	}
}

File: BelajarOOPVBNet/VBNetObjectOP/InheritEncap.vb
Namespace GameOOP
	Public Class Personel
		Private name As String = ""
		Private health As Integer = 0
		Private attackPower As Integer = 0

		Public Property Named As String
			Get
				Return name
			End Get
		    Set(value As String)
				name = value
		    End Set
		End Property

		Public Property Healths As Integer
			Get
				Return health
			End Get
			Set(value As Integer)
				health = If(value >= 0, value, 0)
			End Set
		End Property

		Public Property AttackPowers As Integer
			Get
				Return attackPower
			End Get
			Set(value As Integer)
				attackPower = If(value >= 0, value, 0)
			End Set
		End Property

		Public sub New(name As String, health As Integer, attackPower As Integer)
			Named = name
			Healths = health
			AttackPowers = attackPower
		End sub

		Public Overridable Sub Attack(target As Personel)
			target.Healths -= AttackPowers
			Console.WriteLine($"{target.Named} has {target.Healths} health left.")
		End Sub
	End Class

	Public Class Ranger
		Inherits Personel

		Public Sub New(name As String, health As Integer, attackPower As Integer)
			MyBase.New(name, health, attackPower)
		End Sub

		Public Overrides Sub Attack(target As Personel)
			Console.WriteLine($"{Named} shoots an arrow at {target.Named} for {AttackPowers}")
			MyBase.Attack(target)
		End Sub
	End Class

	Public Class Villain
		Inherits Personel

		Public Sub New(name As String, health As Integer, attackPower As Integer)
			MyBase.New(name, health, attackPower)
		End Sub

		Public Overrides Sub Attack(target As Personel)
			Console.WriteLine($"{Named} casts a spell on {target.Named} for {AttackPowers}")
			MyBase.Attack(target)
		End Sub
	End Class
End Namespace
